Added Module to Support LTECH

Devirex opened the pull request:
* **Please check if the PR fulfills these requirements**
<!-- Please do not remove these checkboxes. Check them if you have done the action behind this point -->
- [x] Tests for the changes have been added / modified (needed for for bug fixes / features)
- [X] commandref has been added / updated (needed for bug fixes / features)


<!--Please specify if this is a bugfix, feature or update of docs  -->
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [ ] Bugfix (please link issue)
- [X] Feature enhancement
- [ ] Documentation update
- [ ] Unittest enhancement
- [ ] other


* **What is the current behavior?** 
No Modul for uknown Protocoll 31

* **What is the new behavior (if this is a feature change)?**
Module LTECH for known Protokol P31 


* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
Module LTECH was added


* **Other information**:

Devirex committed:
Added LTECH

Devirex committed:
Merge branch 'RFD-FHEM:master' into master

Devirex committed:
Debug

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

test

Devirex committed:
delte RAWMSG from def

elektron-bbs committed:
additional sensors protocol 115 for Bresser 6-in-1 Comfort Wetter Center (#1085)

SD_ProtocolData.pm - fixed packet length 18 byte
14_SD_WS.pm - new ground/earth sensor for humidity and temperature, indoor sensor for humidity and temperature

actions-user committed:
Automatic updated controls and CHANGED

sidey79 committed:
Fix wrong calculation of averagewindDirection (#1095)

* Update 14_SD_WS09.pm

- usw CORE Time instead of global from hires
- optimzed code a little bit, removed useless checks
- few perlcritic fixes

actions-user committed:
Automatic updated controls and CHANGED

dependabot[bot] committed:
Bump shogo82148/actions-setup-perl from 1.14.2 to 1.15.1 (#1089)

Bumps [shogo82148/actions-setup-perl](https://github.com/shogo82148/actions-setup-perl) from 1.14.2 to 1.15.1.
- [Release notes](https://github.com/shogo82148/actions-setup-perl/releases)
- [Commits](https://github.com/shogo82148/actions-setup-perl/compare/v1.14.2...v1.15.1)

---
updated-dependencies:
- dependency-name: shogo82148/actions-setup-perl
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: sidey79 <7968127+sidey79@users.noreply.github.com>

dependabot[bot] committed:
Bump actions/upload-artifact from 2 to 3 (#1092)

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 2 to 3.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v2...v3)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: sidey79 <7968127+sidey79@users.noreply.github.com>

dependabot[bot] committed:
Bump actions/download-artifact from 2 to 3 (#1093)

Bumps [actions/download-artifact](https://github.com/actions/download-artifact) from 2 to 3.
- [Release notes](https://github.com/actions/download-artifact/releases)
- [Commits](https://github.com/actions/download-artifact/compare/v2...v3)

---
updated-dependencies:
- dependency-name: actions/download-artifact
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: sidey79 <7968127+sidey79@users.noreply.github.com>

dependabot[bot] committed:
Bump codecov/codecov-action from 2.1.0 to 3.0.0 (#1091)

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 2.1.0 to 3.0.0.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/master/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/v2.1.0...v3.0.0)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: sidey79 <7968127+sidey79@users.noreply.github.com>

elektron-bbs committed:
Protocol 85 new sensor TFA 30.3251.10 (#1096)

14_SD_WS.pm - new sensor TFA 30.3251.10, added CRC8 check for protocol 85

actions-user committed:
Automatic updated controls and CHANGED

Devirex committed:
Added LTECH Module

Devirex committed:
Merge branch 'RFD-FHEM:master' into master

sidey79 commented:
@Devirex 

Danke für den PR.

In dem Modul finden sich noch Reste von einem Siri Modul. Die wolltest Du sicher noch ändern.
Die Perl critic Meldungen beziehen sich hauptsächlich auf die Prototypen aber eine auch wegen der Verwendung von Switch.

Im Detail habe ich mir den Rest jetzt nicht angesehen, grundsätzlich ist es aber so, dass Module unittests benötigen.
Vieles kann in bestehende Tests integriert werden. Manches ist aber speziell. Dafür werden wir Daten von dir benötigten.

Devirex commented:
Wow in der ersten Zeile gleich,
 -> zu lange davor gesessen. Ich passe das an. Das mit den Tests schaue ich mir an.

sidey79 commented:
> Wow in der ersten Zeile gleich, -> zu lange davor gesessen. Ich passe das an. Das mit den Tests schaue ich mir an.

Der einfachste Test prüft nur, ob das Modul geladen werden kann.
Den kannst Du aus einem anderen Modul 1:1 kopieren, dann funktioniert der schon, da er das Modul aus dem Ordnernamen verwendet.
Damit wissen wir dann, dass es keine Syntaxfehler gibt.

Die Parse Funktion wird über Testdaten geprüft.
Das suche sich dir raus, wie das geht.

Einzelne Subs bekommen in der Regel einen eigenen Test. Je nach komplexit der Sub ist das dann einfacher oder aufwändiger.

sidey79 committed:
Merge branch 'master' into master

sidey79 commented:
Also


1) Basistest ob sich das Modul überhaupt laden lässt kannst Du die folgende Datei in einen Ordnernamen identisch zu deinem Modul kopieren. Damit erhalten wir auch gleich die Information ob wir einen Syntaxfehler haben oder nicht:

https://github.com/RFD-FHEM/RFFHEM/blob/master/t/FHEM/10_SD_Rojaflex/00_load.t

In diesem Fall kommen alle Tests in den Ordner `t/FHEM/14_LTECH` 

Damit Tests überhaupt laufen muss auch eine fhem.cfg im Testordner liegen

Diese von hier, könnte ausreichend sein:
https://github.com/RFD-FHEM/RFFHEM/blob/master/t/FHEM/10_SD_GT/fhem.cfg


2a) 
Testdaten kommen hier rein: (JSON)
https://github.com/RFD-FHEM/SIGNALduino_TOOL/blob/master/FHEM/lib/SD_Device_ProtocolList.json#L2975

Sobald hier Testdaten für das Modul enthalten sind, kann ein Test diese validieren. Im groben, werden Internals und Readings deines Moduls damit validiert.

INPUT und OUTPUT sind je nach Test unterschiedlich. z.B. wird validiert ob die RMSG als input zu der angegebenen DMSG führt.
Die DMSG wiederum, wird als Input für andere Tests verwendet um zu validieren, ob ein Reading gesetzt wurde.

Einfach mal versuchen einen Testdatensatz aus RMSG und DMSG zu erstellen.
Lokal innerhalb FHEM kannst Du auch mittels des Moduls [SIGNALduino_TOOL](https://github.com/RFD-FHEM/SIGNALduino_TOOL) testen.


2b)
Damit die Testdaten aus 2a getestet werden, braucht es noch einen weiteren Test.
https://github.com/RFD-FHEM/RFFHEM/blob/master/t/FHEM/14_SD_UT/09_parseDatat.t

Optional können dort auch korrupte Daten getestet werden. Den Teil kommentiere ich aber erst einmal aus


Ich werde dir das gleich mal in deinen PR reinlegen, dann hast Du einen Ausgangspunkt.

sidey79 committed:
t/FHEM/14_LTECH

added base tests for logical module

sidey79 committed:
14_LTECH.pm

- save coderefs as reference and not as text
- save regex as compiled regex instead of string

sidey79 committed:
14_LTECH.pm

Update id / comment / intro

sidey79 committed:
14_LTECH.pm

Update id / comment / intro

Remove Switch Parse

Remove Switch Set

Remove Switch Set

Detect turn off

Detect turn off

Remove Switch import

sidey79 committed:
14_LTECH.pm

Update id / comment / intro

Remove Switch Parse

Remove Switch Set

Remove Switch Set

Detect turn off

Detect turn off

Remove Switch import

sidey79 commented:
@Devirex 

Wie Du siehst, ist der Test nicht erfolgreich, das Modul kann nicht geladen werden und es bricht mit einem Fehler ab:

```
# Failed test '14_LTECH loaded'
Error: # at /home/runner/work/RFFHEM/RFFHEM/t/FHEM/14_LTECH/00_load.t line 13.
# +---------------------------------------------------------+----+---------+
# | GOT                                                     | OP | CHECK   |
# +---------------------------------------------------------+----+---------+
# | Can't locate Switch.pm in @INC (you may need to install | IS | <UNDEF> |
# |  the Switch module) (@INC contains: ./lib ./FHEM . /hom |    |         |
# | e/runner/work/RFFHEM/RFFHEM/local/lib/perl5/5.26.3/x86_ |    |         |
# | 64-linux /home/runner/work/RFFHEM/RFFHEM/local/lib/perl |    |         |
# | 5/5.26.3 /home/runner/work/RFFHEM/RFFHEM/local/lib/perl |    |         |
# | 5/x86_64-linux /home/runner/work/RFFHEM/RFFHEM/local/li |    |         |
# | b/perl5 /home/runner/work/_actions/shogo82148/actions-s |    |         |
# | etup-perl/v1.15.1/scripts/lib /opt/hostedtoolcache/perl |    |         |
# | /5.26.3/x64/lib/site_perl/5.26.3/x86_64-linux /opt/host |    |         |
# | edtoolcache/perl/5.26.3/x64/lib/site_perl/5.26.3 /opt/h |    |         |
# | ostedtoolcache/perl/5.26.3/x64/lib/5.26.3/x86_64-linux  |    |         |
# | /opt/hostedtoolcache/perl/5.26.3/x64/lib/5.26.3) at ./F |    |         |
# | HEM/14_LTECH.pm line 57.\n                              |    |         |
# | BEGIN failed--compilation aborted at ./FHEM/14_LTECH.pm |    |         |
# |  line 57.\n                                             |    |         |
# +---------------------------------------------------------+----+---------+
/home/runner/work/RFFHEM/RFFHEM/t/FHEM/14_LTECH/00_load.t ........................................ 
```

Wie schon erwähnt, sollte switch ersetzt werden.
Da bietet sich eine Lookup Tabelle an um alle Kommandos abzubilden:

```
my %set_commands = (
# cmd switches
rgbcolor => sub { readingsSingleUpdate($hash, 'rgbcolor_sel', uc $value, 1 },
white => sub { readingsSingleUpdate($hash, 'white_sel', $value , 1},
);
```

Anstelle switch wird dann einfach die passende anonyme sub aufgerufen:

`$set_commands {$cmd};`

GGf. müssen aber noch Variable übergeben und eingelesen werden.

Devirex committed:
Merge branch 'devirex-ltech'

Devirex committed:
Added autocreate test

Added autocreate test

Added autocreate test

Added autocreate test

Added autocreate test

Devirex committed:
Added test Data

Devirex committed:
Rename test

Devirex committed:
Added test Data

Devirex committed:
Merge branch 'LTECH-Tests'

Devirex committed:
sub prototype errors

Devirex commented:
> @Devirex
> 
> Wie Du siehst, ist der Test nicht erfolgreich, das Modul kann nicht geladen werden und es bricht mit einem Fehler ab:
> 
> ```
> # Failed test '14_LTECH loaded'
> Error: # at /home/runner/work/RFFHEM/RFFHEM/t/FHEM/14_LTECH/00_load.t line 13.
> # +---------------------------------------------------------+----+---------+
> # | GOT                                                     | OP | CHECK   |
> # +---------------------------------------------------------+----+---------+
> # | Can't locate Switch.pm in @INC (you may need to install | IS | <UNDEF> |
> # |  the Switch module) (@INC contains: ./lib ./FHEM . /hom |    |         |
> # | e/runner/work/RFFHEM/RFFHEM/local/lib/perl5/5.26.3/x86_ |    |         |
> # | 64-linux /home/runner/work/RFFHEM/RFFHEM/local/lib/perl |    |         |
> # | 5/5.26.3 /home/runner/work/RFFHEM/RFFHEM/local/lib/perl |    |         |
> # | 5/x86_64-linux /home/runner/work/RFFHEM/RFFHEM/local/li |    |         |
> # | b/perl5 /home/runner/work/_actions/shogo82148/actions-s |    |         |
> # | etup-perl/v1.15.1/scripts/lib /opt/hostedtoolcache/perl |    |         |
> # | /5.26.3/x64/lib/site_perl/5.26.3/x86_64-linux /opt/host |    |         |
> # | edtoolcache/perl/5.26.3/x64/lib/site_perl/5.26.3 /opt/h |    |         |
> # | ostedtoolcache/perl/5.26.3/x64/lib/5.26.3/x86_64-linux  |    |         |
> # | /opt/hostedtoolcache/perl/5.26.3/x64/lib/5.26.3) at ./F |    |         |
> # | HEM/14_LTECH.pm line 57.\n                              |    |         |
> # | BEGIN failed--compilation aborted at ./FHEM/14_LTECH.pm |    |         |
> # |  line 57.\n                                             |    |         |
> # +---------------------------------------------------------+----+---------+
> /home/runner/work/RFFHEM/RFFHEM/t/FHEM/14_LTECH/00_load.t ........................................ 
> ```
> 
> Wie schon erwähnt, sollte switch ersetzt werden. Da bietet sich eine Lookup Tabelle an um alle Kommandos abzubilden:
> 
> ```
> my %set_commands = (
> # cmd switches
> rgbcolor => sub { readingsSingleUpdate($hash, 'rgbcolor_sel', uc $value, 1 },
> white => sub { readingsSingleUpdate($hash, 'white_sel', $value , 1},
> );
> ```
> 
> Anstelle switch wird dann einfach die passende anonyme sub aufgerufen:
> 
> `$set_commands {$cmd};`
> 
> GGf. müssen aber noch Variable übergeben und eingelesen werden.

Habe jetzt Tests für das Laden, Autocreate und und druchspielen lokaler Testdaten ausprobiert.
Wenn ich alles habe baue ich es dann hier ein:
https://github.com/RFD-FHEM/SIGNALduino_TOOL/blob/master/FHEM/lib/SD_Device_ProtocolList.json#L2975

sidey79 commented on FHEM/14_LTECH.pm:
```diff
@@ -324,7 +324,7 @@ sub Set($@)
             readingsSingleUpdate($hash, 'rgbcolor_sel', $rgbcolor_sel, 1);
             $args[0] = "on"; 
-            Set($hash, $name, @args);
+            &Set($hash, $name, @args);
```
Die Prototypen befinden sich in der Definition, nicht im Aufruf 
Also eher

`sub set($$$) {` durch `sub set {` ersetzen und dann die übergebenen Parameter validieren

z.B.

```
my $hash = shift;
return if ref($hash) ne 'HASH' ;
my $name = shift // return;
...

```

sidey79 commented:
> @sidey79
> 
> Habe jetzt Tests für das Laden, Autocreate und und druchspielen lokaler Testdaten ausprobiert. Wenn ich alles habe baue ich es dann hier ein: https://github.com/RFD-FHEM/SIGNALduino_TOOL/blob/master/FHEM/lib/SD_Device_ProtocolList.json#L2975

Die lokalen Testdaten werden aber nicht geladen, dafür müsstest Du die auskommentierten Zeilen im Test wieder aktivieren :)

Devirex commented:
> 

Also ich hab das gerade 3 Stunden versucht und er ist immer wieder fehlgeschlagen weil ich in den Testdaten vergessen hatte den „

> > @sidey79
> > Habe jetzt Tests für das Laden, Autocreate und und druchspielen lokaler Testdaten ausprobiert. Wenn ich alles habe baue ich es dann hier ein: https://github.com/RFD-FHEM/SIGNALduino_TOOL/blob/master/FHEM/lib/SD_Device_ProtocolList.json#L2975
> 
> Die lokalen Testdaten werden aber nicht geladen, dafür müsstest Du die auskommentierten Zeilen im Test wieder aktivieren :)

Da ist wohl beim merge was schief gelaufen. In meiner Beanch waren die Zeilen nicht auskommentiert :) habe eine Zeit gebraucht um festzustellen dass in meinen Testdaten das DEF bei Internals gefehlt hat und er deshalb das device nicht temporär für den Test erstellt hatte :D

sidey79 commented on t/FHEM/14_LTECH/09_parseData.t:
```diff
@@ -0,0 +1,51 @@
+	my $ioHash = $defs{$ioName};
+    #use Data::Dumper;
+    #print Dumper(@Test::RDmsg::JSONTestList);
+	#my $filepath = dirname(__FILE__);
+	#push @Test2::SIGNALduino::RDmsg::JSONTestList, {
+	#	testname	=> q[Testdata with corrupt SD_UT data],
+	#	url			=> qq[$filepath/testData.json],
+	#};
```
@Devirex 

Das ist der auskommentierte Teil, der lokale Testdaten laden würde.

Devirex committed:
subroutines

Devirex committed:
enable testfile

Devirex committed:
remove debug lines

Devirex committed:
additional tests

Devirex committed:
paramter validation

paramter validation

paramter validation

paramter validation

paramter validation

paramter validation

paramter validation

paramter validation

paramter validation

paramter validation

paramter validation

paramter validation

Devirex committed:
Tests for Set

Tests for Set

Tests for Set

Tests for Set

Tests for Set

Tests for Set

Tests for Set

Tests for Set

Tests for Set

Tests for Set

Tests for Set

sidey79 commented on t/FHEM/14_LTECH/01_define.t:
```diff
@@ -0,0 +1,35 @@
+		plan(1);
+		
+        my $sensorname=q[LTECH_Test_11];
+        CommandDefine(undef,qq[$sensorname LTECH 04444EFF]); 
```
Hier würde ich nur die Define Funktion vom Modul testen und den Rückgabewert prüfen und nicht die Define Funktion von Perl. Außer es gibt einen Grund, den ich jetzt gerade nicht kenne :)

sidey79 commented on t/FHEM/14_LTECH/01_set.t:
```diff
@@ -0,0 +1,202 @@
+
+    subtest 'Protocol 31 - define LTECH_Test_11' => sub {
+		plan(1);
+        CommandDefine(undef,qq[$devicename LTECH 04444EFF]); 
```
Das Gerät, kannst Du auch in fhem.cfg z.B. mit dem Namen LTECH_Test_11 anlegen und dann hier verwenden.
Als Argument wird es ja bereits übergeben.

```
my $name = shift;
my $hash = $defs{$devicename};

```

sidey79 replied:
Würde ich empfehlen.

Devirex committed:
Testdevice in config file

sidey79 committed:
Merge branch 'master' into master

sidey79 committed:
Merge branch 'master' into master

HomeAutoUser committed:
Merge branch 'master' into master

HomeAutoUser commented:
Wie ist hier der Stand @Devirex ?

Devirex commented:
Ist noch was offen? Tests sind drin, config wurde geändert. Benutze das Modul regelmäßig und bin soweit sehr zufrieden.

sidey79 committed:
Merge branch 'master' into master

sidey79 commented:
> Ist noch was offen? Tests sind drin, config wurde geändert. Benutze das Modul regelmäßig und bin soweit sehr zufrieden.

Da sind noch einige Anmerkungen und perlcritic Fehler.
Haben es die Testdaten bereits in das SIGNALDuino Tool geschafft?

HomeAutoUser commented:
> > Ist noch was offen? Tests sind drin, config wurde geändert. Benutze das Modul regelmäßig und bin soweit sehr zufrieden.
> 
> Da sind noch einige Anmerkungen und perlcritic Fehler. Haben es die Testdaten bereits in das SIGNALDuino Tool geschafft?

Ich weiß nicht, ob @Devirex weiß was mit SIGNALDuino Tool gemeint ist.
Aus die Schnelle habe ich nichts gesehen. Wenn ich RAWMSG´s erhalte, so kann ich diese eintragen.

Fehlt sonst noch was hier?

Devirex commented:
Habe einen PR für die Testdaten erstellt:

https://github.com/RFD-FHEM/SIGNALduino_TOOL/pull/76

Devirex committed:
Remove return undef

Devirex committed:
Change color default value

Devirex committed:
Change color default value to match max saturation

sidey79 committed:
Merge branch 'master' into master

sidey79 commented:
Ich habe den Branch aktualisiert und dabei kamen ein paar Abweichungen in den Tests zutage:

```
ok 1 - 14_LTECH loaded
ok
    # Failed test 'check reading for devices'
Error:     # at /home/runner/work/RFFHEM/RFFHEM/t/FHEM/14_LTECH/01_set.t line 105.
    # +--------+----+--------+
    # | GOT    | OP | CHECK  |
    # +--------+----+--------+
    # | 7F1900 | eq | 7F7F7F |
    # +--------+----+--------+
# Failed test 'Protocol 31 - set LTECH_Test_11 brightness 50'
Error: # at /home/runner/work/RFFHEM/RFFHEM/t/FHEM/14_LTECH/01_set.t line 106.
    # Failed test 'check reading for devices'
Error:     # at /home/runner/work/RFFHEM/RFFHEM/t/FHEM/14_LTECH/01_set.t line 139.
    # +--------+----+--------+
    # | GOT    | OP | CHECK  |
    # +--------+----+--------+
    # | 7F1700 | eq | 7F0000 |
    # +--------+----+--------+
# Failed test 'Protocol 31 - set LTECH_Test_11 saturation 100'
Error: # at /home/runner/work/RFFHEM/RFFHEM/t/FHEM/14_LTECH/01_set.t line 140.

```

sidey79 commented on FHEM/14_LTECH.pm:
```diff
@@ -0,0 +1,516 @@
+	my ( $hash, $name, @args ) = @_;
+	my $cmd    = $args[0]; 
+	my $value  = $args[1];
+	Log3( $name, 4, "LTECH: eingehende Werte $cmd , $value");
```
```suggestion
	Log3( $name, 4, "LTECH_Set: eingehende Werte $cmd , $value");
```

Devirex replied:
Übernommen

sidey79 commented on FHEM/14_LTECH.pm:
```diff
@@ -0,0 +1,516 @@
+
+
+	if($cmd eq "rgbcolor" ) { 
+        if($value =~ m/[[:xdigit:]]{6}/){
```
Möglicherweise soll $value ausschließlich aus hexadezimalen Werten bestehen, wenn ja, dann muss noch Start und Ende mit angegeben werden

```suggestion
        if($value =~ m/^[[:xdigit:]]{6}$/){
```

Devirex replied:
Übernommen

sidey79 commented on FHEM/14_LTECH.pm:
```diff
@@ -0,0 +1,516 @@
+my $version = "1.0";
+
+
+sub LTECH_Initialize($) {
```
Die Prototypen sollten überall entfernt werden:

```suggestion
sub LTECH_Initialize {
my $hash= shift // return;
return if ref($hash) ne 'HASH' ;
```

Ich würde es auch anpassen, aber solange die Tests fehl schlagen ändere ich lieber mal nichts.

Devirex replied:
Übernommen

sidey79 commented on FHEM/14_LTECH.pm:
```diff
@@ -0,0 +1,516 @@
+    my @args;
+    my ( $hash, $msg ) = @_;
+	my $name = $hash->{NAME};
+    return "" if ( IsDisabled($name) );	
```
Ich vermute, das macht ich, was Du annimmst.

Was hattest Du denn ursprünglich hiermit vor?

Devirex replied:
War bestandteil des Siro Moduls dass ich las vorlage verwendet hatte. Ist entfernt.

sidey79 commented on FHEM/14_LTECH.pm:
```diff
@@ -0,0 +1,516 @@
+		Log3 $hash, 2, "LTECH: unknown device $deviceCode, please define it";
+        return "UNDEFINED LTECH_$deviceCode LTECH $deviceCode";
+	}
+	$def->{STATE}= "Defined";
```
Unüblich, bist Du sicher, dass du STATE immer mit Defined belegen willst wenn etwas durch Parse verarbeitet wird?

Devirex replied:
Ebenfalls ein überbleibsel aus dem Siro Modul. Ist entfernt.

sidey79 commented on FHEM/14_LTECH.pm:
```diff
@@ -0,0 +1,516 @@
+   my $hash = $defs{$name};
+   my $state = ReadingsVal( $name, 'brightness', 'undef' );
+   my $rgbcolor = ReadingsVal( $name, 'rgbcolor', 'undef' );
+   $hash->{STATE} = $state / 2.55 ;
+   my $sticon = ".*:light_light_dim_". ($state / 2.55)."0\@".$rgbcolor;
+   return $sticon ;
```
wenn in {DEFINED} der Wert "defined" steht, gibt das bestimmt einen Fehler.

Ist das Berechnen eines Icons aber nicht eher etwas für den Anwender über z.B. StateFormat usw. ?

Devirex replied:
Obsolete

Devirex committed:
Merge branch 'RFD-FHEM:master' into master

Devirex committed:
Changes up on recommendation

Devirex committed:
Remove Icon function

sidey79 committed:
Merge branch 'master' into master

sidey79 committed:
Merge branch 'master' into master

HomeAutoUser committed:
Merge branch 'master' into master

HomeAutoUser committed:
Merge branch 'master' into master

sidey79 committed:
Merge branch 'master' into master

HomeAutoUser committed:
Merge branch 'master' into master

Devirex commented:
@HomeAutoUser Ok, wenn von meiner Seite aus noch was fehlt dann bitte Bescheid geben

Devirex committed:
Merge branch 'RFD-FHEM:master' into master

Devirex committed:
fix Module if multiple signalduinos can receive eachother

Devirex committed:
Merge branch 'master' into master

Devirex committed:
Merge branch 'RFD-FHEM:master' into master
